glean-core/csharp/Glean/LibGleanFFI.cs

```diff
@@ -8,6 +8,63 @@
 
 namespace Mozilla.Glean.FFI
 {
+    /// <summary>
+    /// Result values of attempted ping uploads encoded for FFI use.
+    /// They are defined in `glean-core/src/upload/result.rs` and re-defined for use in Kotlin here.
+    /// 
+    /// NOTE:
+    /// THEY MUST BE THE SAME ACROSS BOTH FILES!
+    /// </summary>
+    internal enum Constants : int
+    {
+        // A recoverable error.
+        UPLOAD_RESULT_RECOVERABLE = 0x1,
+
+        // An unrecoverable error.
+        UPLOAD_RESULT_UNRECOVERABLE = 0x2,
+
+        // A HTTP response code.
+        UPLOAD_RESULT_HTTP_STATUS = 0x8000
+    }
+
+    /// <summary>
+    /// Rust represents the upload task as an Enum
+    /// and to go through the FFI that gets transformed into a tagged union.
+    /// Each variant is represented as an 8-bit unsigned integer.
+    ///
+    /// This *MUST* have the same order as the variants in `glean-core/ffi/src/upload.rs`.
+    /// </summary>
+    enum UploadTaskTag : int
+    {
+        Upload,
+        Wait,
+        Done
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct FfiUploadTaskBody
+    {
+        public byte tag;
+        public IntPtr documentId;
+        public IntPtr path;
+        public int bodyLen;
+        public IntPtr body;
+        public IntPtr headers;
+    }
+
+    /// <summary>
+    /// Represent an upload task by simulating the union passed through
+    /// the FFI layer.
+    /// </summary>
+    [StructLayout(LayoutKind.Explicit)]
+    internal struct FfiUploadTask
+    {
+        [FieldOffset(0)]
+        public byte tag;
+        [FieldOffset(0), MarshalAs(UnmanagedType.Struct)]
```

Comment:
Q: I am confused with `FieldOffset(0)` here. It supposes to be `1` for `body` because it shifts one byte from `tag`. But, it looks like you're right. If giving `1`, some tests are skipped.

Reply:
Yes, as you found, the offset is required to be `0` in both places. That's because we're mapping to a [C `union`](https://searchfox.org/glean/rev/73522afaaa64a2c904ec28e209e53c92fb94967b/glean-core/ffi/glean.h#282,289). The `tag` is present both in `FfiUploadTask` and `FfiUploadTaskBody`, at the same location. If tag is "Upload", then we consider this to have a body a proceed with further parsing.